### PR TITLE
feat: Add create ticket feature

### LINKED ISSUE #267: Add Tickets page to show the types of tickets
- Add Tickets page to show each ticket types
- Add ability to add a ticket type

To be done using new API

### PR DESCRIPTION
**Pull request by iamareebjamal**
Fixes #267 

- Add ticket POST methods in service and repository
- Add ticket database change listener
- Create view and presenter for creating tickets
- Add ticket binder for binding layout to ticket fields
- Add create ticket FAB in tickets page
- Fix format for ISO date in DateUtils
- Add base dialog fragment
- Add appropriate icons
- Fix event serialization by ignoring it
- Make ticket fields class type in order to ignore nulls
- Create intermediate field for observing creation

screenshot-1501061872745
screenshot-1501061881288

### CODE REVIEW

**Comment by mejariamol** [comment]
Attendee creation should be preferred before this. What do you say? @iamareebjamal

**Comment by iamareebjamal** [comment]
Attendee creation is independent of this. Furthermore, in a new Event, you cannot create an attendee without a ticket. And attendee without order too can't exist and order API is not ready yet

**Comment by iamareebjamal** [comment]
@mejariamol Please review

**Comment by mejariamol** [comment]
@iamareebjamal please have a look at this error. App is crashing on start. If not try re-installing the app. Not sure if it is from this pr. picasso provider is returning null I guess.
```org.fossasia.openevent.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.fossasia.openevent.app, PID: 6737
    java.lang.RuntimeException: Unable to create application org.fossasia.openevent.app.OrgaApplication: java.lang.NullPointerException: Cannot return null from a non-@Nullable @Provides method
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4752)
        at android.app.ActivityThread.access$1500(ActivityThread.java:162)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1473)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:189)
        at android.app.ActivityThread.main(ActivityThread.java:5530)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:950)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:745)
    Caused by: java.lang.NullPointerException: Cannot return null from a non-@Nullable @Provides method
        at dagger.internal.Preconditions.checkNotNull(Preconditions.java:48)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_TokenFactory.get(NetworkModule_TokenFactory.java:27)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_TokenFactory.get(NetworkModule_TokenFactory.java:9)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_AuthenticatorFactory.get(NetworkModule_AuthenticatorFactory.java:28)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_AuthenticatorFactory.get(NetworkModule_AuthenticatorFactory.java:9)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_ProvidesOkHttpClientFactory.get(NetworkModule_ProvidesOkHttpClientFactory.java:53)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_ProvidesOkHttpClientFactory.get(NetworkModule_ProvidesOkHttpClientFactory.java:13)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_ProvidesPicassoFactory.get(NetworkModule_ProvidesPicassoFactory.java:37)
        at org.fossasia.openevent.app.common.app.di.module.NetworkModule_ProvidesPicassoFactory.get(NetworkModule_ProvidesPicassoFactory.java:11)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at org.fossasia.openevent.app.OrgaApplication_MembersInjector.injectMembers(OrgaApplication_MembersInjector.java:29)
        at org.fossasia.openevent.app.OrgaApplication_MembersInjector.injectMembers(OrgaApplication_MembersInjector.java:8)
        at org.fossasia.openevent.app.common.app.di.component.DaggerAppComponent.inject(DaggerAppComponent.java:556)
        at org.fossasia.openevent.app.OrgaApplication.onCreate(OrgaApplication.java:83)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1011)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4749)
        at android.app.ActivityThread.access$1500(ActivityThread.java:162) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1473) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:189) 
        at android.app.ActivityThread.main(ActivityThread.java:5530) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:950) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:745) 
```

**Comment by iamareebjamal** [comment]
No, this is token issue. I thought it'd be good to provide token as a dependency. It'll simplify the network module. Seems that @Provides can't return null. Solving this

**Comment by iamareebjamal** [comment]
@mejariamol Please test now

**Comment by mejariamol** [comment]
```
Error:(22, 10) error: java.lang.String is not nullable, but is being provided by @Provides @javax.annotation.Nullable @Named("token") String org.fossasia.openevent.app.common.app.di.module.NetworkModule.token(org.fossasia.openevent.app.common.data.contract.IUtilModel)
at:     @javax.inject.Named("token") java.lang.String is injected at
org.fossasia.openevent.app.common.app.di.module.NetworkModule.authenticator(token)
okhttp3.Authenticator is injected at
org.fossasia.openevent.app.common.app.di.module.NetworkModule.providesOkHttpClient(…, authenticator)
okhttp3.OkHttpClient is injected at
org.fossasia.openevent.app.common.app.di.module.NetworkModule.providesRetrofitBuilder(…, client)
retrofit2.Retrofit.Builder is injected at
org.fossasia.openevent.app.common.app.di.module.NetworkModule.providesEventService(retrofitBuilder)
org.fossasia.openevent.app.common.data.network.EventService is injected at
org.fossasia.openevent.app.common.data.repository.AttendeeRepository.<init>(…, eventService)
org.fossasia.openevent.app.common.data.repository.AttendeeRepository is injected at
org.fossasia.openevent.app.common.app.di.module.DataModule.providesAttendeeRepository(attendeeRepository)
org.fossasia.openevent.app.common.data.repository.contract.IAttendeeRepository is injected at
org.fossasia.openevent.app.module.attendee.qrscan.ScanQRPresenter.<init>(attendeeRepository)
org.fossasia.openevent.app.module.attendee.qrscan.ScanQRPresenter is injected at
org.fossasia.openevent.app.common.app.di.module.PresenterModule.bindsScanQRPresenter(scanQRPresenter)
dagger.Lazy<org.fossasia.openevent.app.module.attendee.qrscan.contract.IScanQRPresenter> is injected at
org.fossasia.openevent.app.module.attendee.qrscan.ScanQRActivity.presenterProvider
org.fossasia.openevent.app.module.attendee.qrscan.ScanQRActivity is injected at
org.fossasia.openevent.app.common.app.di.component.BarcodeComponent.inject(scanQRActivity)
```
@iamareebjamal

File: app/src/main/java/org/fossasia/openevent/app/common/app/di/module/NetworkModule.java [hunk 1]
```diff
@@ -93,8 +94,15 @@ StethoInterceptor stethoInterceptor() {
...
+    @Nullable
+    @Named("token")
+    String token(IUtilModel utilModel) {
+        return utilModel.getToken();
+    }
+
+    @Provides
     @Singleton
-    Authenticator authenticator(IUtilModel utilModel) {
+    Authenticator authenticator(@Named("token") String token) {
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/common/app/di/module/NetworkModule.java]: You may need to mark nullable to the token here too.

**Comment by iamareebjamal** [comment]
@mejariamol Thanks. Updated

**Comment by mejariamol** [comment]
@iamareebjamal making token a dependency is creating an issue. I am getting 401 on authorization required requests. please verify if the token is updated in Authenticator on change.

**Comment by iamareebjamal** [comment]
Did you log out and log in?

**Comment by iamareebjamal** [comment]
@mejariamol I've restored the old DI model

**Comment by mejariamol** [comment]
@iamareebjamal some modifications are required:
1. make ticket name text view single line as next button  not working on it
2. on back pressed, hide the soft keyboard if shown or just close the ticket create dialog. Currently on back pressed the app is closed.
3. I am getting 422 error now on creating ticket.
4. price text view should be hidden if free option is chosen.

**Comment by iamareebjamal** [comment]
1. Next Button will not work even after single line is added as the next field is radio group
2. onBackPressed is overriden by the activity and not a fragment. And this should be handled by a different PR
3. Because the sales start and ends are same time. Increase the time of end to be more than start. There will be date pickers and validators for this, just not in this PR
4. This is also to be added in another PR

The goal of issues for creating features should be to make bare minimum working and bug-free solution, and further enhancements should be added later. This already took 3 days, adding more things would make it complicated. Let's test if the feature is working and does not break the things currently working and move on to enhance it :smile:

**Comment by mejariamol** [comment]
@iamareebjamal 
>onBackPressed is overriden by the activity and not a fragment. And this should be handled by a different PR

but it is working fine in the current master branch

**Comment by iamareebjamal** [comment]
@mejariamol The double tap feature is working fine in this branch too

**Comment by mejariamol** [comment]
Yeah. Sorry. I have updated my comment. :smile: 
Don't know why it was not working in earlier test.

**Comment by iamareebjamal** [comment]
@niranjan94 Please merge

**Comment by mariobehling** [comment]
Please resolve conflicts.

**Pull request merged**